src/fundamental/macro-fundamental.h

```diff
@@ -249,6 +249,30 @@
                         CONST_ISPOWEROF2(_x);                          \
                 }))
 
+#define ADD_SAFE(ret, a, b) (!__builtin_add_overflow((a), (b), (ret)))
+#define INC_SAFE(a, b) __INC_SAFE(UNIQ, (a), (b))
+#define __INC_SAFE(q, a, b)                                     \
+        ({                                                      \
+                const typeof(a) UNIQ_T(A, q) = (a);             \
+                ADD_SAFE(UNIQ_T(A, q), *UNIQ_T(A, q), (b));     \
+        })
+
+#define SUB_SAFE(ret, a, b) (!__builtin_sub_overflow((a), (b), (ret)))
+#define DEC_SAFE(a, b) __DEC_SAFE(UNIQ, (a), (b))
+#define __DEC_SAFE(q, a, b)                                     \
+        ({                                                      \
+                const typeof(a) UNIQ_T(A, q) = (a);             \
+                SUB_SAFE(UNIQ_T(A, q), *UNIQ_T(A, q), (b));     \
+        })
+
+#define MUL_SAFE(ret, a, b) (!__builtin_mul_overflow((a), (b), (ret)))
+#define MUL_ASSIGN_SAFE(a, b) __MUL_ASSIGN_SAFE(UNIQ, (a), (b))
+#define __MUL_ASSIGN_SAFE(q, a, b)                              \
+        ({                                                      \
+                const typeof(a) UNIQ_T(A, q) = (a);             \
+                MUL_SAFE(UNIQ_T(A, q), *UNIQ_T(A, q), (b));     \
+        })
```

Comment:
i guess these helpers make sense, but can you add a quick set of tests to src/test/test-macro.c or so?